File: src/mnel/placement.py
"""Backend-neutral accelerator placement policy for learned providers.

The policy deliberately does not import Torch, Accelerate, or a vendor runtime. A
backend supplies capability diagnostics and an adapter applies the returned decision.
This keeps deterministic policy tests cheap and keeps the Rust host architecture
independent from any one accelerator stack.
"""

from __future__ import annotations

from dataclasses import asdict, dataclass
from enum import StrEnum
from typing import Any

MIB = 1024 * 1024
DEFAULT_GPU_RESERVE_MIB = 256
DEFAULT_WORKSPACE_MIB = 256


class PlacementError(RuntimeError):
    """The requested placement cannot be satisfied safely."""


class ExecutionDevice(StrEnum):
    AUTO = "auto"
    CPU = "cpu"
    CUDA = "cuda"


class OffloadMode(StrEnum):
    AUTO = "auto"
    NONE = "none"
    SEQUENTIAL_CPU = "sequential-cpu"


class Precision(StrEnum):
    AUTO = "auto"
    FLOAT32 = "float32"
    FLOAT16 = "float16"
    BFLOAT16 = "bfloat16"


class ExecutionMode(StrEnum):
    CPU = "cpu"
    FULL_CUDA = "full-cuda"
    SEQUENTIAL_CPU_OFFLOAD = "sequential-cpu-offload"


@dataclass(frozen=True, slots=True)
class PlacementPolicy:
    execution_device: ExecutionDevice = ExecutionDevice.AUTO
    offload: OffloadMode = OffloadMode.AUTO
    precision: Precision = Precision.AUTO
    gpu_reserve_mib: int = DEFAULT_GPU_RESERVE_MIB
    max_vram_mib: int | None = None
    model_storage_bytes: int = 0
    workspace_bytes: int = DEFAULT_WORKSPACE_MIB * MIB
    host_memory_budget_bytes: int | None = None

    def validate(self) -> None:
        if self.execution_device is ExecutionDevice.CPU and self.offload is OffloadMode.SEQUENTIAL_CPU:
            raise PlacementError("sequential CPU offload requires auto or cuda execution device")
        if self.gpu_reserve_mib < 0:
            raise PlacementError("GPU reserve cannot be negative")
        if self.max_vram_mib is not None and self.max_vram_mib < 1:
            raise PlacementError("maximum VRAM must be positive")
        if self.model_storage_bytes < 0 or self.workspace_bytes < 0:
            raise PlacementError("model and workspace estimates cannot be negative")
        if self.host_memory_budget_bytes is not None and self.host_memory_budget_bytes < 1:
            raise PlacementError("host memory budget must be positive")


@dataclass(frozen=True, slots=True)
class PlacementCapabilities:
    supports_sequential_cpu_offload: bool = False
    cpu_precisions: tuple[Precision, ...] = (Precision.FLOAT32,)
    cuda_precisions: tuple[Precision, ...] = (
        Precision.FLOAT32,
        Precision.FLOAT16,
        Precision.BFLOAT16,
    )

    def validate(self) -> None:
        if not self.cpu_precisions:
            raise PlacementError("backend must declare at least one CPU precision")
        if not self.cuda_precisions:
            raise PlacementError("backend must declare at least one CUDA precision")


@dataclass(frozen=True, slots=True)
class AcceleratorDiagnostics:
    accelerator_available: bool = False
    execution_probe_succeeded: bool = False
    free_vram_bytes: int | None = None
    accelerator_identity: str | None = None
    float16_probe_succeeded: bool | None = None
    bfloat16_probe_succeeded: bool | None = None
    probe_error: str | None = None

    def as_dict(self) -> dict[str, Any]:
        return asdict(self)


@dataclass(frozen=True, slots=True)
class PlacementDecision:
    execution_mode: ExecutionMode
    execution_device: ExecutionDevice
    offload: OffloadMode
    precision: Precision
    reason: str
    configured_gpu_reserve_bytes: int
    configured_max_vram_bytes: int | None
    effective_gpu_budget_bytes: int
    estimated_model_bytes: int
    estimated_workspace_bytes: int
    full_cuda_required_bytes: int
    host_memory_required_bytes: int
    sequential_offload_supported: bool

    def as_dict(self) -> dict[str, Any]:
        value = asdict(self)
        for key in ("execution_mode", "execution_device", "offload", "precision"):
            value[key] = value[key].value
        return value


def effective_gpu_budget_bytes(
    free_vram_bytes: int | None,
    reserve_mib: int,
    max_vram_mib: int | None,
) -> int:
    """Return free VRAM remaining after the operator reserve and optional cap."""

    if free_vram_bytes is None:
        return 0
    capped = free_vram_bytes
    if max_vram_mib is not None:
        capped = min(capped, max_vram_mib * MIB)
    return max(0, capped - reserve_mib * MIB)


def _precision_bytes(model_storage_bytes: int, precision: Precision) -> int:
    if precision in {Precision.FLOAT16, Precision.BFLOAT16}:
        return (model_storage_bytes + 1) // 2
    return model_storage_bytes


def _choose_cuda_precision(
    policy: PlacementPolicy,
    diagnostics: AcceleratorDiagnostics,
    capabilities: PlacementCapabilities,
) -> Precision:
    if policy.precision is not Precision.AUTO:
        if policy.precision not in capabilities.cuda_precisions:
            raise PlacementError(f"CUDA backend does not declare {policy.precision.value} support")
        if policy.precision is Precision.FLOAT16 and diagnostics.float16_probe_succeeded is not True:
            raise PlacementError("float16 was requested but its execution probe failed")
        if policy.precision is Precision.BFLOAT16 and diagnostics.bfloat16_probe_succeeded is not True:
            raise PlacementError("bfloat16 was requested but its execution probe failed")
        return policy.precision
    if (
        Precision.BFLOAT16 in capabilities.cuda_precisions
        and diagnostics.bfloat16_probe_succeeded is True
    ):
        return Precision.BFLOAT16
    if Precision.FLOAT16 in capabilities.cuda_precisions and diagnostics.float16_probe_succeeded is True:
        return Precision.FLOAT16
    if Precision.FLOAT32 in capabilities.cuda_precisions:
        return Precision.FLOAT32
    raise PlacementError("CUDA backend has no usable precision")


def decide_placement(
    policy: PlacementPolicy,
    diagnostics: AcceleratorDiagnostics,
    capabilities: PlacementCapabilities,
) -> PlacementDecision:
    """Choose CPU, full CUDA, or true sequential CPU offload deterministically."""

    policy.validate()
    capabilities.validate()
    model_bytes = policy.model_storage_bytes
    host_budget = policy.host_memory_budget_bytes
    if host_budget is not None and model_bytes > host_budget:
        raise PlacementError("provider model exceeds the host/system-memory budget")

    reserve_bytes = policy.gpu_reserve_mib * MIB
    maximum_bytes = policy.max_vram_mib * MIB if policy.max_vram_mib is not None else None
    gpu_budget = effective_gpu_budget_bytes(
        diagnostics.free_vram_bytes, policy.gpu_reserve_mib, policy.max_vram_mib
    )
    cuda_usable = diagnostics.accelerator_available and diagnostics.execution_probe_succeeded

    def decision(
        mode: ExecutionMode,
        precision: Precision,
        reason: str,
    ) -> PlacementDecision:
        estimated_model = _precision_bytes(model_bytes, precision)
        return PlacementDecision(
            execution_mode=mode,
            execution_device=(
                ExecutionDevice.CUDA if mode is not ExecutionMode.CPU else ExecutionDevice.CPU
            ),
            offload=(
                OffloadMode.SEQUENTIAL_CPU
                if mode is ExecutionMode.SEQUENTIAL_CPU_OFFLOAD
                else OffloadMode.NONE
            ),
            precision=precision,
            reason=reason,
            configured_gpu_reserve_bytes=reserve_bytes,
            configured_max_vram_bytes=maximum_bytes,
            effective_gpu_budget_bytes=gpu_budget,
            estimated_model_bytes=estimated_model,
            estimated_workspace_bytes=policy.workspace_bytes,
            full_cuda_required_bytes=estimated_model + policy.workspace_bytes,
            host_memory_required_bytes=model_bytes,
            sequential_offload_supported=capabilities.supports_sequential_cpu_offload,
        )

    if policy.execution_device is ExecutionDevice.CPU:
        if policy.precision is Precision.AUTO:
            precision = Precision.FLOAT32
        else:
            precision = policy.precision
        if precision not in capabilities.cpu_precisions:
            raise PlacementError(f"CPU backend does not declare {precision.value} support")
        return decision(ExecutionMode.CPU, precision, "CPU was explicitly requested")

    if not cuda_usable:
        reason = diagnostics.probe_error or "accelerator execution probe failed"
        if (
            policy.execution_device is ExecutionDevice.CUDA
            or policy.offload is OffloadMode.SEQUENTIAL_CPU
        ):
            raise PlacementError(f"CUDA execution is unusable: {reason}")
        return decision(ExecutionMode.CPU, Precision.FLOAT32, f"AUTO selected CPU because {reason}")

    precision = _choose_cuda_precision(policy, diagnostics, capabilities)
    required = _precision_bytes(model_bytes, precision) + policy.workspace_bytes
    fits = required <= gpu_budget

    if policy.offload is OffloadMode.SEQUENTIAL_CPU:
        if not capabilities.supports_sequential_cpu_offload:
            raise PlacementError("sequential CPU offload is unsupported by this provider")
        return decision(
            ExecutionMode.SEQUENTIAL_CPU_OFFLOAD,
            precision,
            "sequential CPU offload was explicitly requested",
        )

    if policy.offload is OffloadMode.NONE:
        if not fits:
            if policy.execution_device is ExecutionDevice.AUTO:
                return decision(
                    ExecutionMode.CPU,
                    Precision.FLOAT32,
                    "AUTO selected CPU because full CUDA exceeds the effective budget and offload is disabled",
                )
            raise PlacementError("full CUDA exceeds the effective GPU budget")
        return decision(
            ExecutionMode.FULL_CUDA,
            precision,
            "full CUDA was explicitly requested" if policy.execution_device is ExecutionDevice.CUDA else "full CUDA fits the effective GPU budget",
        )

    if fits:
        return decision(ExecutionMode.FULL_CUDA, precision, "full CUDA fits the effective GPU budget")
    if capabilities.supports_sequential_cpu_offload:
        return decision(
            ExecutionMode.SEQUENTIAL_CPU_OFFLOAD,
            precision,
            "full CUDA exceeds the effective GPU budget; using CPU-backed sequential execution",
        )
    if policy.execution_device is ExecutionDevice.AUTO:
        return decision(
            ExecutionMode.CPU,
            Precision.FLOAT32,
            "AUTO selected CPU because full CUDA exceeds budget and provider offload is unsupported",
        )
    raise PlacementError(
        "full CUDA exceeds the effective GPU budget and sequential offload is unsupported"
    )


def fallback_policy_after_oom(
    policy: PlacementPolicy,
    current_mode: ExecutionMode,
    capabilities: PlacementCapabilities,
) -> PlacementPolicy | None:
    """Return one bounded AUTO retry policy; explicit operator choices never retry."""

    if policy.execution_device is not ExecutionDevice.AUTO:
        return None
    if current_mode is ExecutionMode.FULL_CUDA and capabilities.supports_sequential_cpu_offload:
        return PlacementPolicy(
            execution_device=ExecutionDevice.AUTO,
            offload=OffloadMode.SEQUENTIAL_CPU,
            precision=policy.precision,
            gpu_reserve_mib=policy.gpu_reserve_mib,
            max_vram_mib=policy.max_vram_mib,
            model_storage_bytes=policy.model_storage_bytes,
            workspace_bytes=policy.workspace_bytes,
            host_memory_budget_bytes=policy.host_memory_budget_bytes,
        )
    if current_mode in {ExecutionMode.FULL_CUDA, ExecutionMode.SEQUENTIAL_CPU_OFFLOAD}:
        return PlacementPolicy(
            execution_device=ExecutionDevice.CPU,
            offload=OffloadMode.NONE,
            precision=Precision.FLOAT32,
            model_storage_bytes=policy.model_storage_bytes,
            workspace_bytes=policy.workspace_bytes,
            host_memory_budget_bytes=policy.host_memory_budget_bytes,
        )
    return None

File: crates/mnel-provider-host/src/placement.rs
//! Backend-neutral accelerator placement policy.
//!
//! This module records policy and decisions only. A CUDA, vendor, or Torch adapter
//! supplies diagnostics and applies the decision behind an explicit capability boundary.

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum ExecutionDevice {
    Auto,
    Cpu,
    Cuda,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum OffloadMode {
    Auto,
    None,
    SequentialCpu,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum Precision {
    Auto,
    Float32,
    Float16,
    Bfloat16,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum ExecutionMode {
    Cpu,
    FullCuda,
    SequentialCpuOffload,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct PlacementPolicy {
    pub execution_device: ExecutionDevice,
    pub offload: OffloadMode,
    pub precision: Precision,
    pub gpu_reserve_bytes: u64,
    pub max_vram_bytes: Option<u64>,
    pub model_storage_bytes: u64,
    pub workspace_bytes: u64,
    pub host_memory_budget_bytes: Option<u64>,
}

impl Default for PlacementPolicy {
    fn default() -> Self {
        Self {
            execution_device: ExecutionDevice::Auto,
            offload: OffloadMode::Auto,
            precision: Precision::Auto,
            gpu_reserve_bytes: 256 * 1024 * 1024,
            max_vram_bytes: None,
            model_storage_bytes: 0,
            workspace_bytes: 256 * 1024 * 1024,
            host_memory_budget_bytes: None,
        }
    }
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct PlacementCapabilities {
    pub supports_sequential_cpu_offload: bool,
    pub cpu_precisions: Vec<Precision>,
    pub cuda_precisions: Vec<Precision>,
}

impl Default for PlacementCapabilities {
    fn default() -> Self {
        Self {
            supports_sequential_cpu_offload: false,
            cpu_precisions: vec![Precision::Float32],
            cuda_precisions: vec![Precision::Float32, Precision::Float16, Precision::Bfloat16],
        }
    }
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct AcceleratorDiagnostics {
    pub accelerator_available: bool,
    pub execution_probe_succeeded: bool,
    pub free_vram_bytes: Option<u64>,
    pub accelerator_identity: Option<String>,
    pub float16_probe_succeeded: Option<bool>,
    pub bfloat16_probe_succeeded: Option<bool>,
    pub probe_error: Option<String>,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct PlacementDecision {
    pub execution_mode: ExecutionMode,
    pub execution_device: ExecutionDevice,
    pub offload: OffloadMode,
    pub precision: Precision,
    pub reason: String,
    pub configured_gpu_reserve_bytes: u64,
    pub configured_max_vram_bytes: Option<u64>,
    pub effective_gpu_budget_bytes: u64,
    pub estimated_model_bytes: u64,
    pub estimated_workspace_bytes: u64,
    pub full_cuda_required_bytes: u64,
    pub host_memory_required_bytes: u64,
    pub sequential_offload_supported: bool,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum PlacementError {
    InvalidPolicy,
    MissingCpuPrecision,
    MissingCudaPrecision,
    HostMemoryBudgetExceeded,
    CudaUnavailable,
    UnsupportedPrecision,
    PrecisionProbeFailed,
    SequentialOffloadUnsupported,
    FullCudaBudgetExceeded,
}

fn precision_bytes(model_storage_bytes: u64, precision: Precision) -> u64 {
    match precision {
        Precision::Float16 | Precision::Bfloat16 => model_storage_bytes.div_ceil(2),
        Precision::Auto | Precision::Float32 => model_storage_bytes,
    }
}

fn cuda_precision(
    policy: &PlacementPolicy,
    diagnostics: &AcceleratorDiagnostics,
    capabilities: &PlacementCapabilities,
) -> Result<Precision, PlacementError> {
    let precision = match policy.precision {
        Precision::Auto => {
            if capabilities.cuda_precisions.contains(&Precision::Bfloat16)
                && diagnostics.bfloat16_probe_succeeded == Some(true)
            {
                Precision::Bfloat16
            } else if capabilities.cuda_precisions.contains(&Precision::Float16)
                && diagnostics.float16_probe_succeeded == Some(true)
            {
                Precision::Float16
            } else {
                Precision::Float32
            }
        }
        requested => requested,
    };
    if !capabilities.cuda_precisions.contains(&precision) || precision == Precision::Auto {
        return Err(PlacementError::UnsupportedPrecision);
    }
    if (precision == Precision::Float16 && diagnostics.float16_probe_succeeded != Some(true))
        || (precision == Precision::Bfloat16 && diagnostics.bfloat16_probe_succeeded != Some(true))
    {
        return Err(PlacementError::PrecisionProbeFailed);
    }
    Ok(precision)
}

pub fn effective_gpu_budget_bytes(
    free_vram_bytes: Option<u64>,
    reserve_bytes: u64,
    max_vram_bytes: Option<u64>,
) -> u64 {
    let Some(free) = free_vram_bytes else {
        return 0;
    };
    let capped = max_vram_bytes.map_or(free, |maximum| free.min(maximum));
    capped.saturating_sub(reserve_bytes)
}

pub fn decide_placement(
    policy: &PlacementPolicy,
    diagnostics: &AcceleratorDiagnostics,
    capabilities: &PlacementCapabilities,
) -> Result<PlacementDecision, PlacementError> {
    if policy.execution_device == ExecutionDevice::Cpu
        && policy.offload == OffloadMode::SequentialCpu
    {
        return Err(PlacementError::InvalidPolicy);
    }
    if capabilities.cpu_precisions.is_empty() {
        return Err(PlacementError::MissingCpuPrecision);
    }
    if capabilities.cuda_precisions.is_empty() {
        return Err(PlacementError::MissingCudaPrecision);
    }
    if policy.max_vram_bytes == Some(0) || policy.host_memory_budget_bytes == Some(0) {
        return Err(PlacementError::InvalidPolicy);
    }
    if policy
        .host_memory_budget_bytes
        .is_some_and(|budget| policy.model_storage_bytes > budget)
    {
        return Err(PlacementError::HostMemoryBudgetExceeded);
    }

    let gpu_budget = effective_gpu_budget_bytes(
        diagnostics.free_vram_bytes,
        policy.gpu_reserve_bytes,
        policy.max_vram_bytes,
    );
    let usable = diagnostics.accelerator_available && diagnostics.execution_probe_succeeded;

    let make_decision = |mode: ExecutionMode, precision: Precision, reason: &str| {
        let estimated_model = precision_bytes(policy.model_storage_bytes, precision);
        PlacementDecision {
            execution_mode: mode,
            execution_device: if mode == ExecutionMode::Cpu {
                ExecutionDevice::Cpu
            } else {
                ExecutionDevice::Cuda
            },
            offload: if mode == ExecutionMode::SequentialCpuOffload {
                OffloadMode::SequentialCpu
            } else {
                OffloadMode::None
            },
            precision,
            reason: reason.to_owned(),
            configured_gpu_reserve_bytes: policy.gpu_reserve_bytes,
            configured_max_vram_bytes: policy.max_vram_bytes,
            effective_gpu_budget_bytes: gpu_budget,
            estimated_model_bytes: estimated_model,
            estimated_workspace_bytes: policy.workspace_bytes,
            full_cuda_required_bytes: estimated_model.saturating_add(policy.workspace_bytes),
            host_memory_required_bytes: policy.model_storage_bytes,
            sequential_offload_supported: capabilities.supports_sequential_cpu_offload,
        }
    };

    if policy.execution_device == ExecutionDevice::Cpu {
        let precision = if policy.precision == Precision::Auto {
            Precision::Float32
        } else {
            policy.precision
        };
        if !capabilities.cpu_precisions.contains(&precision) {
            return Err(PlacementError::UnsupportedPrecision);
        }
        return Ok(make_decision(
            ExecutionMode::Cpu,
            precision,
            "CPU was explicitly requested",
        ));
    }

    if !usable {
        if policy.execution_device == ExecutionDevice::Cuda
            || policy.offload == OffloadMode::SequentialCpu
        {
            return Err(PlacementError::CudaUnavailable);
        }
        return Ok(make_decision(
            ExecutionMode::Cpu,
            Precision::Float32,
            "AUTO selected CPU because accelerator execution is unavailable",
        ));
    }

    let precision = cuda_precision(policy, diagnostics, capabilities)?;
    let required = precision_bytes(policy.model_storage_bytes, precision)
        .saturating_add(policy.workspace_bytes);
    let fits = required <= gpu_budget;
    if policy.offload == OffloadMode::SequentialCpu {
        if !capabilities.supports_sequential_cpu_offload {
            return Err(PlacementError::SequentialOffloadUnsupported);
        }
        return Ok(make_decision(
            ExecutionMode::SequentialCpuOffload,
            precision,
            "sequential CPU offload was explicitly requested",
        ));
    }
    if policy.offload == OffloadMode::None {
        if !fits {
            if policy.execution_device == ExecutionDevice::Auto {
                return Ok(make_decision(
                    ExecutionMode::Cpu,
                    Precision::Float32,
                    "AUTO selected CPU because full CUDA exceeds budget and offload is disabled",
                ));
            }
            return Err(PlacementError::FullCudaBudgetExceeded);
        }
        return Ok(make_decision(
            ExecutionMode::FullCuda,
            precision,
            "full CUDA fits the effective GPU budget",
        ));
    }
    if fits {
        return Ok(make_decision(
            ExecutionMode::FullCuda,
            precision,
            "full CUDA fits the effective GPU budget",
        ));
    }
    if capabilities.supports_sequential_cpu_offload {
        return Ok(make_decision(
            ExecutionMode::SequentialCpuOffload,
            precision,
            "full CUDA exceeds budget; using CPU-backed sequential execution",
        ));
    }
    if policy.execution_device == ExecutionDevice::Auto {
        return Ok(make_decision(
            ExecutionMode::Cpu,
            Precision::Float32,
            "AUTO selected CPU because full CUDA exceeds budget and offload is unsupported",
        ));
    }
    Err(PlacementError::FullCudaBudgetExceeded)
}

pub fn fallback_after_oom(
    policy: &PlacementPolicy,
    current_mode: ExecutionMode,
    capabilities: &PlacementCapabilities,
) -> Option<PlacementPolicy> {
    if policy.execution_device != ExecutionDevice::Auto {
        return None;
    }
    if current_mode == ExecutionMode::FullCuda && capabilities.supports_sequential_cpu_offload {
        return Some(PlacementPolicy {
            execution_device: ExecutionDevice::Auto,
            offload: OffloadMode::SequentialCpu,
            ..policy.clone()
        });
    }
    if matches!(
        current_mode,
        ExecutionMode::FullCuda | ExecutionMode::SequentialCpuOffload
    ) {
        return Some(PlacementPolicy {
            execution_device: ExecutionDevice::Cpu,
            offload: OffloadMode::None,
            precision: Precision::Float32,
            ..policy.clone()
        });
    }
    None
}

#[cfg(test)]
mod tests {
    use super::*;

    fn cuda(free_mib: u64) -> AcceleratorDiagnostics {
        AcceleratorDiagnostics {
            accelerator_available: true,
            execution_probe_succeeded: true,
            free_vram_bytes: Some(free_mib * 1024 * 1024),
            accelerator_identity: Some("fake-cuda".to_owned()),
            float16_probe_succeeded: Some(true),
            bfloat16_probe_succeeded: Some(false),
            probe_error: None,
        }
    }

    #[test]
    fn reserve_and_cap_math_is_deterministic() {
        assert_eq!(
            effective_gpu_budget_bytes(
                Some(2048 * 1024 * 1024),
                256 * 1024 * 1024,
                Some(1024 * 1024 * 1024)
            ),
            768 * 1024 * 1024
        );
    }

    #[test]
    fn auto_selects_sequential_offload_when_full_cuda_does_not_fit() {
        let policy = PlacementPolicy {
            model_storage_bytes: 2048 * 1024 * 1024,
            workspace_bytes: 512 * 1024 * 1024,
            ..PlacementPolicy::default()
        };
        let capabilities = PlacementCapabilities {
            supports_sequential_cpu_offload: true,
            ..PlacementCapabilities::default()
        };
        let decision = match decide_placement(&policy, &cuda(1024), &capabilities) {
            Ok(decision) => decision,
            Err(_) => panic!("placement should succeed"),
        };
        assert_eq!(decision.execution_mode, ExecutionMode::SequentialCpuOffload);
    }

    #[test]
    fn explicit_cuda_budget_failure_is_not_silently_recovered() {
        let policy = PlacementPolicy {
            execution_device: ExecutionDevice::Cuda,
            offload: OffloadMode::None,
            model_storage_bytes: 2048 * 1024 * 1024,
            ..PlacementPolicy::default()
        };
        assert_eq!(
            decide_placement(&policy, &cuda(1024), &PlacementCapabilities::default()),
            Err(PlacementError::FullCudaBudgetExceeded)
        );
    }
}
